---
title: "Transcriptome Demo"
author: "Jair"
date: "`r Sys.Date()`"
output: pdf_document
---

# Load required packages
```{r message=FALSE}
library(ballgown) 
library(RColorBrewer)
library(genefilter)
library(dplyr)
library(devtools)
```

#Organizes data into a dataframe that has two columns.
#The first column contains the ids while the second specifies the corresponding stage, either planktonic or biofilm.
```{r}
pheno_data<-data.frame(ids = c("plank01", "plank02", "biofilm01", "biofilm02"),
                       stage = c("planktonic", "planktonic", "biofilm", "biofilm"))
```

#Creates a Ballgown object and check transcript number
#Organizes the file path into a vector
```{r message=FALSE}
samples.c <- paste('ballgown', pheno_data$ids, sep = '/')
bg <- ballgown(samples = samples.c, meas='all', pData = pheno_data)
bg
```

#Filters out transcripts with a small variance.
#Only transcripts with a variance larger than 1 are kept.
```{r}
bg_filt = subset(bg,"rowVars(texpr(bg)) >1",genomesubset=TRUE)
bg_filt
```

#Creates a table of transcripts
```{r}
results_transcripts<- stattest(bg_filt, feature = "transcript", covariate = "stage",
getFC = TRUE, meas = "FPKM")
results_transcripts<-data.frame(geneNames=geneNames(bg_filt),
transcriptNames=transcriptNames(bg_filt), results_transcripts)
```

#Shows the top of the table stored in results_transcripts
```{r}
head(results_transcripts)
```

#Choose a transcript to examine more closely
```{r}
results_transcripts[results_transcripts$transcriptNames == "gene-PA0004", ] #I chose to examine gene-PA0004 (gyrB)
```
##This transcript is for the gene gyrB. 
##The fold difference between planktonic and biofilm stages is 48342981060
##The corrected q-value of 0.9471885 indicates that the difference is not significant.

#Filters out non-significant transcripts.
#Only results with a significant p-value (less than 0.05) are kept
#The dim function gives the number of transcripts kept.
```{r}
sigdiff <- results_transcripts %>% filter(pval<0.05)
dim(sigdiff)
```

#Organizes the table by p-value and fold change (fc).
#The p-values go from smallest to biggest (since decreasing=FALSE)
#The fold change is in increasing order since the negative sign in front and setting decreasing=FALSE.
```{r}
o = order(sigdiff[,"pval"], -abs(sigdiff[,"fc"]), decreasing=FALSE)
output = sigdiff[o,c("geneNames","transcriptNames", "id","fc","pval","qval")]
write.table(output, file="SigDiff.txt", sep="\t", row.names=FALSE, quote=FALSE)
head(output)
```

#Loads gene names
```{r}
bg_table = texpr(bg_filt, 'all')
bg_gene_names = unique(bg_table[, 9:10])
```

#Pulls out gene expression data and visualize
```{r}
gene_expression = as.data.frame(gexpr(bg_filt))
head(gene_expression)
```

#Renames the columns to the names specified within c("")
```{r}
colnames(gene_expression) <- c("plank01", "plank02", "biofilm01", "biofilm02")
head(gene_expression)
dim(gene_expression)
```

#Loads the transcript to gene table and determine the number of transcripts and unique genes
```{r}
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)
length(row.names(transcript_gene_table))
length(unique(transcript_gene_table[,"g_id"]))
```

#Plots the number of transcripts per gene
```{r}
counts=table(transcript_gene_table[,"g_id"])
c_one = length(which(counts == 1))
c_more_than_one = length(which(counts > 1))
c_max = max(counts)
hist(counts, breaks=50, col="bisque4", xlab="Transcripts per gene",
main="Distribution of transcript count per gene")
legend_text = c(paste("Genes with one transcript =", c_one),
paste("Genes with more than one transcript =", c_more_than_one),
paste("Max transcripts for single gene = ", c_max))
legend("topright", legend_text, lty=NULL)
```

##A vast majority of the genes have a single transcript.
##Few genes have multpiple transcripts
##The most transcripts for a single gene were 64.

#Creates a plot of how similar the two planktonic replicates are to one another.
```{r}
x = gene_expression[,"plank01"]
y = gene_expression[,"plank02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="purple", cex=0.25,
xlab="FPKM (Planktonic, Replicate 1)", ylab="FPKM (Planktonic, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```

#Creates a plot of how similar the two biofilm replicates are to one another.
```{r}
x = gene_expression[,"biofilm01"]
y = gene_expression[,"biofilm02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="blue", cex=0.25,
xlab="FPKM (Biofilm, Replicate 1)", ylab="FPKM (Biofilm, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```
##Both replicates within each condition were similar.
##If both replicates within a single condition are similar, this means we can reliable compare between conditions (planktonic and biofilm)

#Creates plot of differential gene expression between the conditions
```{r}
results_genes = stattest(bg_filt, feature="gene", covariate="stage", getFC=TRUE, meas="FPKM")
results_genes = merge(results_genes,bg_gene_names,by.x=c("id"),by.y=c("gene_id"))
sig=which(results_genes$pval<0.05)
results_genes[,"de"] = log2(results_genes[,"fc"])
hist(results_genes[sig,"de"], breaks=50, col="seagreen",
xlab="log2(Fold change) Planktonic vs Biofilm",
main="Distribution of differential expression values")
abline(v=-2, col="black", lwd=2, lty=2)
abline(v=2, col="black", lwd=2, lty=2)
legend("topleft", "Fold-change > 4", lwd=2, lty=2)
```
##The differential gene expression plot is bimodal, indicating a difference in gene expression between both conditions.
#Many genes had a fold-change greater than 4, suggesting a large difference in expression base on the condition (planktonic or biofilm)

#Plots total gene expression highlighting differentially expressed genes
```{r}
gene_expression[,"plank"]=apply(gene_expression[,c(1:2)], 1, mean)
gene_expression[,"biofilm"]=apply(gene_expression[,c(3:4)], 1, mean)
x=log2(gene_expression[,"plank"]+min_nonzero)
y=log2(gene_expression[,"biofilm"]+min_nonzero)
plot(x=x, y=y, pch=16, cex=0.25, xlab="Planktonic FPKM (log2)", ylab="Biofilm FPKM (log2)",
main="Planktonic vs Biofilm FPKMs")
abline(a=0, b=1)
xsig=x[sig]
ysig=y[sig]
points(x=xsig, y=ysig, col="magenta", pch=16, cex=0.5)
legend("topleft", "Significant", col="magenta", pch=16)
```

#Makes a table of FPKM values
```{r}
fpkm = texpr(bg_filt,meas="FPKM")
```


#Chooses a gene to determine individual expression
```{r}
ballgown::transcriptNames(bg_filt)[10]
ballgown::geneNames(bg_filt)[10]
```

#Transforms to log2
```{r}
transformed_fpkm <- log2(fpkm[2, ] + 1)
```

#Makes sure values are properly coded as numbers
```{r}
numeric_stages <- as.numeric(factor(pheno_data$stage))

jittered_stages <- jitter(numeric_stages)
```

#Plots expression of individual gene
```{r}
boxplot(transformed_fpkm ~ pheno_data$stage,
        main=paste(ballgown::geneNames(bg_filt)[10], ' : ', ballgown::transcriptNames(bg_filt)[10]),
        xlab="Stage",
        ylab="log2(FPKM+1)",
        col=c("lightblue", "salmon"),
        border="darkblue")

points(transformed_fpkm ~ jittered_stages, 
       pch=21, col="blue", bg="lightblue", cex=1.2)
```
##gene-PA0010 had a higher expression in the planktonic than in the biofilm stage.