---
title: "Microbiome DADA2"
author: "Jair"
date: "2024-04-16"
output: pdf_document
---

# Load required packages
```{r}
library(dada2)
```

# Load sequences
```{r}
path <- "sequences"
list.files(path)
```

# fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
```{r}
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

# Inspect read quality
```{r}
plotQualityProfile(fnFs[1:2])
```

```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
names(filtFs) <- sample.names

#According to our QualityProfile, the quality decreases at 120
out <- filterAndTrim(fnFs, filtFs, truncLen=c(120),
              maxN=0, maxEE=c(2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE)
head(out)
```

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
plotErrors(errF, nominalQ=TRUE)
```

# Sample inference
```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
```

# Amplicon sequence variant table (ASV) table (819 ASVs detected)
```{r}
seqtab <- makeSequenceTable(dadaFs)
dim(seqtab)
```


```{r}
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

#Removal of chimeras
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
```



```{r}
sum(seqtab.nochim)/sum(seqtab)
```

# Tracking reads through the pipeline
```{r}
getN <- function(x) sum(getUniques(x))
# Only use dada results for forward reads, simplify if only one sample
if(length(dadaFs) == 1) {
    denoisedF <- getN(dadaFs)
} else {
    denoisedF <- sapply(dadaFs, getN)
}

# Assemble the tracking matrix
track <- cbind(out, denoisedF, rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoisedF", "nonchim")
rownames(track) <- sample.names
head(track)
```
# Assigning taxa using the Silva reference database
```{r}
taxa <- assignTaxonomy(seqtab.nochim, "silva_nr99_v138.1_train_set.fa", multithread=TRUE)
```

```{r}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```

```{r}
save(taxa, file = "RData/taxa.RData")
```

#save setab.nochim as an R file
```{r}
if (!dir.exists("RData")) {
  dir.create("RData")
}
save(seqtab.nochim, file = "RData/seqtab.nochim.RData")

```
